geomesa-kafka/geomesa-kafka-datastore/src/test/scala/org/locationtech/geomesa/kafka/data/KafkaDataStoreTest.scala

```diff
@@ -525,6 +526,66 @@ class KafkaDataStoreTest extends Specification with Mockito with LazyLogging {
       }
     }
 
+    "support at-least-once consumers" >> {
+      val params = Map(
+        KafkaDataStoreParams.ConsumerConfig.key -> "auto.offset.reset=earliest",
+        KafkaDataStoreParams.ConsumerCount.key -> "2",
+        KafkaDataStoreParams.TopicPartitions.key -> "2"
+      )
+      val (producer, consumer, sft) = createStorePair("batch-consumers", params)
+      try {
+        val id = "fid-0"
+        val numUpdates = 3
+        val maxLon = 80.0
+
+        val seen = new AtomicBoolean(false)
+        var results = new CopyOnWriteArrayList[SimpleFeature]().asScala
+
+        val processor = new GeoMessageProcessor() {
+          override def consume(records: Seq[GeoMessage]): Boolean = {
+            val features = records.collect { case GeoMessage.Change(f) => f }
+            if (!seen.get) {
+              seen.set(true)
+              false // this should cause the messages to be replayed
+            } else {
+              results ++= features
+              true
+            }
+          }
+        }
+
+        producer.createSchema(sft)
+
+        def writeUpdates(): Unit = {
+          WithClose(producer.getFeatureWriterAppend(sft.getTypeName, Transaction.AUTO_COMMIT)) { writer =>
+            (numUpdates to 1 by -1).foreach { i =>
+              val ll = maxLon - maxLon / i
+              val sf = writer.next()
+              sf.setAttributes(Array[AnyRef]("smith", Int.box(30), new Date(), s"POINT ($ll $ll)"))
+              sf.getIdentifier.asInstanceOf[FeatureIdImpl].setID(s"$id-$ll")
+              sf.getUserData.put(Hints.USE_PROVIDED_FID, java.lang.Boolean.TRUE)
+              writer.write()
+            }
+          }
+        }
+
+        WithClose(consumer.createConsumer(sft.getTypeName, "mygroup", processor)) { _ =>
+          writeUpdates()
+          eventually(seen.get must beTrue)
+          eventually(results must haveLength(numUpdates))
+        }
+
+        // verify that we can read a second batch
+        writeUpdates()
+        WithClose(consumer.createConsumer(sft.getTypeName, "mygroup", processor)) { _ =>
+          eventually(results must haveLength(numUpdates * 2))
```

Comment:
Is this showing that the records were processed twice?

Reply:
it's verifying that the second batch of records was also read - we write one batch, read the messages, close the consumer, write a second batch, create a new consumer with the same group id and verify that it picked up from where it left off.